src/test/file.rs

```diff
@@ -158,3 +960,664 @@ pub fn load(filename: &str) -> Result<test::File, Box<dyn Error + Send + Sync>>
+                    .get(rng.gen_range(0..vals.len()))
+                    .unwrap_or(&T::default())
+                    .clone(),
+                None => rng.gen_range(
+                    spec.min.clone().unwrap_or_default()..spec.max.clone().unwrap_or_default(),
+                ),
+            };
+        })
+        .filter(|v| {
+            spec.check(v, &|_e, _a| "".to_string())
+                .into_iter()
+                .collect::<Validated<Vec<()>, String>>()
+                .is_good()
+        })
+        .nth(0)
+}
+
+pub fn generate_float(spec: &FloatSpecification, max_attemps: u16) -> Option<f64> {
+    if spec.value.is_some() {
+        return spec.value.clone();
+    }
+
+    let mut rng = rand::thread_rng();
+    (0..max_attemps)
+        .map(|_| {
+            return match spec.one_of.as_ref() {
+                Some(vals) => vals
+                    .get(rng.gen_range(0..vals.len()))
+                    .unwrap_or(&f64::default())
+                    .clone(),
+                None => rng.gen_range(
+                    spec.min.clone().unwrap_or_default()..spec.max.clone().unwrap_or_default(),
+                ),
+            };
+        })
+        .filter(|v| {
+            spec.check(v, &|_e, _a| "".to_string())
+                .into_iter()
+                .collect::<Validated<Vec<()>, String>>()
+                .is_good()
+        })
+        .nth(0)
+}
+
+pub fn generate_string(spec: &Specification<String>, max_attemps: u16) -> Option<String> {
+    const CHARSET: &[u8] = b"ABCDEFGHIJKLMNOPQRSTUVWXYZ\
+                            abcdefghijklmnopqrstuvwxyz\
+                            "; // 0123456789)(*&^%$#@!~";
+
+    if spec.value.is_some() {
+        return spec.value.clone();
+    }
+
+    let mut rng = rand::thread_rng();
+    let string_length: usize = rng.gen_range(1..50);
+
+    for _ in 0..max_attemps {
+        let ret: String = match spec.one_of.as_ref() {
+            Some(vals) => vals
+                .get(rng.gen_range(0..vals.len()))
+                .unwrap_or(&String::default())
+                .clone(),
+            None => (0..string_length)
+                .map(|_| {
+                    let idx = rng.gen_range(0..CHARSET.len());
+                    CHARSET[idx] as char
+                })
+                .collect::<String>(),
+        };
+
+        let r = spec.check(&ret, &|_e, _a| "".to_string());
+        if r.into_iter()
+            .collect::<Validated<Vec<()>, String>>()
+            .is_good()
+        {
+            return Some(ret);
+        }
+    }
+
+    None
+}
+
+pub fn generate_date(spec: &DateSpecification, max_attemps: u16) -> Option<String> {
+    if let Some(val) = &spec.specification.value {
+        return spec.get(val).ok();
+    }
+
+    //What we need for randomness below
+    //let mut rng = rand::thread_rng();
+    //let string_length: usize = rng.gen_range(DateTime::UNIX_EPOCH..DateTime::naive_local());
+
+    //Have to check it
```

Comment:
for date generation should we not just determine the min/max year, month, day, etc. start with year and randomly generate integers in the range it's allowed, and then narrow the range of the others as we go? I assume max_attempts is because each cycle may not produce a "correct" value.

Reply:
Yes, this looks incomplete. Let me address that.